src/main/kotlin/org/jetbrains/research/testspark/helpers/psi/kotlin/KotlinPsiClassWrapper.kt

```diff
@@ -0,0 +1,125 @@
+import org.jetbrains.kotlin.idea.refactoring.isInterfaceClass
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.KtClass
+import org.jetbrains.kotlin.psi.KtClassOrObject
+import org.jetbrains.kotlin.psi.KtObjectDeclaration
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.DescriptorToSourceUtils
+import org.jetbrains.research.testspark.core.data.ClassType
+import org.jetbrains.research.testspark.core.utils.importPattern
+import org.jetbrains.research.testspark.core.utils.packagePattern
+import org.jetbrains.research.testspark.helpers.psi.PsiClassWrapper
+import org.jetbrains.research.testspark.helpers.psi.PsiMethodWrapper
+
+class KotlinPsiClassWrapper(private val psiClass: KtClassOrObject) : PsiClassWrapper {
+    override val name: String get() = psiClass.name ?: ""
+
+    override val qualifiedName: String get() = psiClass.fqName?.asString() ?: ""
+
+    override val text: String? get() = psiClass.text
+
+    override val methods: List<PsiMethodWrapper>
+        get() = psiClass.body?.functions?.map { KotlinPsiMethodWrapper(it) } ?: emptyList()
+
+    override val allMethods: List<PsiMethodWrapper> get() = methods
+
+    override val superClass: PsiClassWrapper?
+        get() {
+            // Get the superTypeListEntries of the Kotlin class
+            val superTypeListEntries = psiClass.superTypeListEntries
+            // Find the superclass entry (if any)
+            val superClassEntry = superTypeListEntries.firstOrNull()
+            // Resolve the superclass type reference to a PsiClass
+            val superClassTypeReference = superClassEntry?.typeReference
+            val superClassDescriptor = superClassTypeReference?.let {
+                val bindingContext = it.analyze()
+                bindingContext[BindingContext.TYPE, it]
+            }
+            val superClassPsiClass = superClassDescriptor?.constructor?.declarationDescriptor?.let { descriptor ->
+                DescriptorToSourceUtils.getSourceFromDescriptor(descriptor) as? KtClass
+            }
+            return superClassPsiClass?.let { KotlinPsiClassWrapper(it) }
+        }
+
+    override val virtualFile: VirtualFile get() = psiClass.containingFile.virtualFile
+
+    override val containingFile: PsiFile get() = psiClass.containingFile
+
+    override val fullText: String
+        get() {
+            var fullText = ""
+            val fileText = psiClass.containingFile.text
+
+            // get package
+            packagePattern.findAll(fileText, 0).map {
+                it.groupValues[0]
+            }.forEach {
+                fullText += "$it\n\n"
+            }
+
+            // get imports
+            importPattern.findAll(fileText, 0).map {
+                it.groupValues[0]
+            }.forEach {
+                fullText += "$it\n"
+            }
+
+            // Add class code
+            fullText += psiClass.text
+
+            return fullText
+        }
+
```

Comment:
It seams to be a complete copy-paste of the implementation within [JavaPsiClassWrapper](https://github.com/JetBrains-Research/TestSpark/blob/development/src/main/kotlin/org/jetbrains/research/testspark/helpers/psi/java/JavaPsiClassWrapper.kt). 

It is better to either have it in the base class (but in this case, issues might arise once we move to support other languages _[maybe, idk]_) or create a separate function.

Reply:
for another languages this function will probably look differently

Reply:
Still, I wouldn't say I like the idea of having it as a duplicate. 

We are encountering an [OOP problem with the inheritance hierarchies (StackOverflow)](https://stackoverflow.com/a/10715053). The answer suggests to use the **Strategy** pattern here to resolve it, although it introduces a boilerplate:

```kotlin
// strategy interface for the extract of a full text of a class
interface ClassFullTextExtractionStratery {
	fun extract(containingFile: VirtualFile, classText: String): String
}

// direct implementor for the Java and Kotlin PsiWrappers
JavaKotlinFullTextExtractionStratery : ClassFullTextExtractionStratery {
	override fun extract(containingFile: VirtualFile, classText: String): String {
		var fullText = ""
        val fileText = containingFile.text

        // get package
        packagePattern.findAll(fileText, 0).map {
            it.groupValues[0]
        }.forEach {
            fullText += "$it\n\n"
        }

        // get imports
        importPattern.findAll(fileText, 0).map {
            it.groupValues[0]
        }.forEach {
            fullText += "$it\n"
        }

        // Add class code
        fullText += classText

        return fullText
	}
}

// use case for Kotlin
class KotlinPsiClassWrapper {
	// ...
	val textExtractor: ClassFullTextExtractionStratery = JavaKotlinFullTextExtractionStratery()

	val fullText: String
		get() = textExtractor.extract(psiClass.containingFile, psiClass.text)
}

// use case for Java
class JavaPsiClassWrapper {
	// ...
	val textExtractor: ClassFullTextExtractionStratery = JavaKotlinFullTextExtractionStratery()

	val fullText: String
		get() = textExtractor.extract(psiClass.containingFile, psiClass.text)
}
```


I personally find the solution viable, but it is a rather intrusive change, so it is better to negotiate it @pderakhshanfar, @arksap2002. If we opt to introduce this change, it should be done in a different PR.

Reply:
Obviously, avoiding boilerplate is more favorable. However, this is a refactoring for another PR.

Reply:
I will introduce this change in the different PR 
[An issue](https://github.com/JetBrains-Research/TestSpark/issues/265) is opened.